airsonic-main/src/main/resources/liquibase/legacy/schema25.xml

```diff
@@ -92,21 +92,20 @@
         </rollback>
     </changeSet>
     <changeSet id="schema25_005" author="muff1nman">
-        <validCheckSum>8:4d92e7c97abc6f197f1236b0353ec2f7</validCheckSum>
+        <validCheckSum>8:73d2b294465ee42488a8f0238cf686cc</validCheckSum>
         <preConditions onFail="MARK_RAN">
             <not>
                 <tableExists tableName="user_role" />
             </not>
         </preConditions>
         <createTable tableName="user_role">
             <column name="username" type="${varchar_type}">
-                <constraints nullable="false" foreignKeyName="ur_u_fk" referencedTableName="user" referencedColumnNames="username"/>
+                <constraints primaryKey="true" foreignKeyName="ur_u_fk" referencedTableName="user" referencedColumnNames="username"/>
```

Comment:
These should all be done in new migrations. Editing existing ones is a last resort.

Reply:
Spring Boot 2 is producing an incorrect SQL for these migrations. These migrations have to be changed. Adding another migration later won't change the incorrect SQL generated here that is persisted in the HSQLDB script file.

Please see the output of the failed tests for https://github.com/airsonic/airsonic/pull/1372 which shows the incorrect SQL this migration is generating. That PR won't pass until this is merged into _master_.
```
Error creating bean with name 'liquibase' defined in class path resource [org/airsonic/player/spring/DatabaseConfiguration.class]: Invocation of init method failed; nested exception is liquibase.exception.DatabaseException: java.sql.SQLException: error in script file line: 7 Constraint already exists: UR_U_FK in statement [CREATE MEMORY TABLE USER_ROLE(USERNAME VARCHAR NOT NULL,ROLE_ID INTEGER NOT NULL,CONSTRAINT UR_U_FK PRIMARY KEY(USERNAME,ROLE_ID),CONSTRAINT UR_U_FK FOREIGN KEY(USERNAME) REFERENCES "USER"(USERNAME),CONSTRAINT UR_R_FK FOREIGN KEY(ROLE_ID) REFERENCES "ROLE"(ID))]
```

Notice the name of the primary key is the same as the name of a foreign key (UR_U_FK), thus the duplicate constraint error

Reply:
Any idea things are going to be different if a column is defined as a primary key in some user databases and not others? I'd rather guarantee that everyone has the exact same db schema, even if we have to somehow handle this manually in the migrations.

Reply:
Sorry, not sure I understand the question? Mind clarifying?

This migration retains the same primary keys, so nothing should change. It's just "written" differently in liquibase parlance.

Previously: You add columns and then impose a primary key constraint on those columns in the next line (addPrimaryKey tag)
Now: You add columns and specify they're part of a primary key as part of the column declaration itself (and remove the unnecessary addPrimaryKey tag)

Either way, the same columns end up being primary keys in SQL.

From a migration perspective:
- People with new DBs: won't know the difference. The new migrations get applied and the same columns (as the older migrations) end up being primary keys.
- People with old DBs: no-ops. The table already exists with the same primary keys, the migrations don't even need to run. Only thing that changes are the validation checksums stored by liquibase for the migrations, which we've specified.

Reply:
Thanks for the detailed explanation, that makes it clearer! I think there aren't any issues with the PR then.